### PR TITLE
Unit Test - Merkle

### PR DESCRIPTION
**Pull request by bh2smith**
- Four test cases giving nearly full coverage of Merkle.sol
- Included a few changes to Plasma.sol that were mostly lint-related.

### CODE REVIEW

File: contracts/mocks/MerkleMock.sol [hunk 1]
```diff
@@ -0,0 +1,18 @@
+pragma solidity ^0.4.24;
+
+import "../utils/Merkle.sol";
+
+
+contract MerkleMock {
```
Comment (fleupold) [inline contracts/mocks/MerkleMock.sol]: Would we be able to use the mock-contract here? Right now it doesn't really seem to be a mock but instead do some actual computation.

File: test/merkle.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  })
+
+  describe('Containment proofs', function () {
+
+    it('Height 3, leaf 0, proof 0', async () => {
+      var leaf = '0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d'
+      var index = 0
+      var tree_root = '0x568ff5eb286f51b8a3e8de4e53aa8daed44594a246deebbde119ea2eb27acd6b'
+      var proof = '0xc89efdaa54c0f20c7adf612882df0950f5a951637e0307cdcb4c672f298b8bc627c2feed9df4c4903bfc9f1bda886662bebc8ba175ccdb3d1da20ce3a32441ba519d76fbbd63a3ac72cdfee0c0e650d76e933ef34227d72ef272a397ee3ce5ba'
+      res = (await merkle.checkMembership(leaf, index, tree_root, proof, 3));
```
Comment (fleupold) [inline test/merkle.js]: It looks like you are testing the Mock here. The class under test should really never be mocked. Mocks should be injected into the class under test instead of real life dependencies (e.g. a ERC20 token that is passed in the constructor).


Why do you need this extra object? Would it make sense to add that convenience function to the actual contract itself?

File: test/merkle.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    merkle = await MerkleMock.new();
+  })
+
+  describe('Containment proofs', function () {
+
+    it('Height 3, leaf 0, proof 0', async () => {
+      var leaf = '0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d'
+      var index = 0
+      var tree_root = '0x568ff5eb286f51b8a3e8de4e53aa8daed44594a246deebbde119ea2eb27acd6b'
+      var proof = '0xc89efdaa54c0f20c7adf612882df0950f5a951637e0307cdcb4c672f298b8bc627c2feed9df4c4903bfc9f1bda886662bebc8ba175ccdb3d1da20ce3a32441ba519d76fbbd63a3ac72cdfee0c0e650d76e933ef34227d72ef272a397ee3ce5ba'
```
Comment (fleupold) [inline test/merkle.js]: It would be awesome if you could somehow programatically derive this value. Tests can also serve as a great documentation on how the library is expected to be used. Right now I have no idea how you come up with this value.

File: test/merkle.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+contract('Merkle', (accounts) => {
+
+  before(async () => {
+    merkle = await MerkleMock.new();
+  })
+
+  describe('Containment proofs', function () {
+
+    it('Height 3, leaf 0, proof 0', async () => {
```
Comment (fleupold) [inline test/merkle.js]: What does proof 0 mean?

File: test/merkle.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      assert.equal(res, true, "Successful submission of proof height");
+    })
+
+    it('Height 3, bad index', async () => {
+      var leaf = '0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d'
+      var index = 1
+      var tree_root = '0x568ff5eb286f51b8a3e8de4e53aa8daed44594a246deebbde119ea2eb27acd6b'
+      var proof = '0xc89efdaa54c0f20c7adf612882df0950f5a951637e0307cdcb4c672f298b8bc627c2feed9df4c4903bfc9f1bda886662bebc8ba175ccdb3d1da20ce3a32441ba519d76fbbd63a3ac72cdfee0c0e650d76e933ef34227d72ef272a397ee3ce5ba'
+      res = (await merkle.checkMembership(leaf, index, tree_root, proof, 3));
+      assert.equal(res, false, "Not contained! Good proof, bad index");
```
Comment (fleupold) [inline test/merkle.js]: nit: Maybe say "index doesn't match proof"?

**Review by fleupold — CHANGES_REQUESTED** [review]
I believe you don't want to use a Mock here, but instead expose `checkMembership` with the `height` parameter somehow in the test. I'm usually not a fan of subclassing things just for tests, as unit tests should really just test the public interface of a class and not implementation details.
Therefore, if we want to test the sub-function, I'd argue we should expose it publicly on the main contract. Currently it's made internal and I don't even see from where it is used.

How long does the computation of a single merkle proof take in the test? Would it be feasible to test all permutation of a let's say 4 element tree + a bad index and a bad hash? 

Could we also add tests for if `proof` is too long or too short or `index` is out of range?

File: test/merkle.js (hunk 2 above)
Comment (bh2smith) [reply test/merkle.js]: This is apparently the standard for testing libraries (i.e. they need to be used by a contract). I gathered this knowledge  from the openzeppelin-solidity repo as this is how they test their librares.  [Here](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/mocks/MathMock.sol) is an example of a contract using *Math.sol*. 

Considering that Merkle should be a utility contract and potentially moved into a standalone location, I also think that it should have its own unit tests. It is, after all, a standalone structure.

File: test/merkle.js (hunk 3 above)
Comment (bh2smith) [reply test/merkle.js]: Yes I agree completely. I constructed these examples in python, could you suggest a nice way to include these pieces into the project? Perhaps a "generator" directory inside test? Would this interfere with truffle's test suite?

File: test/merkle.js (hunk 4 above)
Comment (bh2smith) [reply test/merkle.js]: It means it is the proof for the leaf at index zero (i.e. correct proof). I will try to elaborate these better. Suggestions warmly welcome.

File: contracts/mocks/MerkleMock.sol (hunk 1 above)
Comment (bh2smith) [reply contracts/mocks/MerkleMock.sol]: It doesn't appear to be a mock with direct regards to [your](https://github.com/fleupold/mock-contract) your notion, but it does coincide with the unit tests for solidity libraries according to [openzeppelin](https://github.com/OpenZeppelin/openzeppelin-solidity) Are there any other suggested terms one might want to consider?

File: test/merkle.js (hunk 5 above)
Comment (bh2smith) [reply test/merkle.js]: Yes, actually I should change this completely since this message is only meant to be displayed upon failure of passing the test case.

File: test/merkle.js (hunk 3 above)
Comment (fleupold) [reply test/merkle.js]: How about just generating them with js code right in the test?

File: test/merkle.js (hunk 2 above)
Comment (fleupold) [reply test/merkle.js]: Ah I get it now. I missed that Merkle.sol is a library and not a contract. Sorry about the confusion.

I think the name Mock is very misleading and something that OpenZeppelin should consider revisiting (I'm happy to open an issue in their repo). A mock is something that should never be the system under test.
It's really just a `MerkleFacadeContract` or `MerkleWrapperContract` for the underlying library.

I'd also argue all it should do it call the library method directly and not have any additional if/else logic.

File: contracts/mocks/MerkleMock.sol (hunk 1 above)
Comment (fleupold) [reply contracts/mocks/MerkleMock.sol]: as per suggestion below. Maybe MerkleLibraryFacade or MerkleLibraryWrapper?

File: test/merkle.js (hunk 3 above)
Comment (bh2smith) [reply test/merkle.js]: Ya, I'm sure that could be done... might take me sometime to get it right.

File: test/merkle.js (hunk 2 above)
Comment (bh2smith) [reply test/merkle.js]: I was trying to call the library methods directly, but got some dumb complaint about not specifying the right number of parameters for my method call. In order to work around this, I used the if statement to define only a single function that called both. 

Would you agree that we don't actually need those two duplicate methods in the Merkle Library? I would be very happy to take the one without the "height" parameter out. Its not like typing 16 every time is an inconvenience. Suggestions? 

I would be happy to land on an appropriate alternative such as "Facade" or "Wrapper".

File: test/merkle.js (hunk 3 above)
Comment (josojo) [reply test/merkle.js]: Let's try do to it. We need it anyways later for other tests

File: test/merkle.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
+const MerkleMock = artifacts.require("MerkleMock");
+
+const MerkleTree = require('merkletreejs');
+const { sha3 } = require('ethereumjs-util')
+
+let {
+    assertRejects,
+    catchError,
+    toHex,
+    fastForward,
```
Comment (josojo) [inline test/merkle.js]: Do you need this one? What is the proof ForDepositBlock for?

**Review by josojo — APPROVED** [review]
Like it a lot.

File: test/merkle.js [hunk 7]
```diff
@@ -1,51 +1,109 @@
...
+    proofForDepositBlock,
+    zeroHashes
+} = require('./utilities.js');
+
+
 contract('Merkle', (accounts) => {
 
   before(async () => {
     merkle = await MerkleMock.new();
```
Comment (fleupold) [inline test/merkle.js]: I'd prefer doing this in each test case, to avoid state being stored between test cases. It's not really important here since we are not storing any state, but I'd prefer to use this pattern generally.

File: test/merkle.js [hunk 8]
```diff
@@ -1,51 +1,109 @@
...
+      await assertRejects(merkle.checkMembership(hex_leaves[1], 1, hex_root, hex_proofs[1], 5));
+    })
+
+    it('Wrong Proof length; 0-0-0', async () => {
+      bad_proof = hex_proofs[0].slice(0, 10)
+      await assertRejects(merkle.checkMembership(hex_leaves[0], 0, hex_root, bad_proof, 2));
+    })
+
+  })
+describe('Height 4: check left and right most txn', function () {
```
Comment (fleupold) [inline test/merkle.js]: Is this test adding any value, given that you have already tested height 2? If at all should we check height 16 or does this take too long?

File: test/utilities.js [hunk 9]
```diff
@@ -4,6 +4,22 @@ let RLP = require('rlp');
  How to avoid using try/catch blocks with promises' that could fail using async/await
  - https://blog.grossman.io/how-to-write-async-await-without-try-catch-blocks-in-javascript/
  */
+
+const assertRejects = async (q, msg) => {
```
Comment (fleupold) [inline test/utilities.js:8]: I'm surprised this code isn't super common. Can we maybe import this from somewhere instead of copying the code in our repo?

File: test/merkle.js (hunk 6 above)
Comment (fleupold) [reply test/merkle.js]: Same with catchError, zeroHashes

**Review by fleupold — COMMENTED** [review]
Looks good now. Just a few minor suggestions. I'd like to see the rename happening before accepting.

File: test/merkle.js (hunk 3 above)
Comment (bh2smith) [reply test/merkle.js]: All done with the JS-tests! That was fun!

File: test/plasma.js [hunk 10]
```diff
@@ -8,7 +8,7 @@ var etherToken
 var plasma
 
 // const c1 = () =>
-contract('Plasma - DepositTests', (accounts) => {
+contract('Plasma', (accounts) => {
```
Comment (fleupold) [inline test/plasma.js]: nit: these changes are not related to this pull request.

File: test/merkle.js [hunk 11]
```diff
@@ -16,7 +16,7 @@ let {
 contract('Merkle', (accounts) => {
 
   before(async () => {
-    merkle = await MerkleMock.new();
+    merkle = await MerkleWrapper.new();
```
Comment (fleupold) [inline test/merkle.js]: as per previous comment, I'd prefer if we create a new object under test inside each test to avoid dependencies on the order of execution.

**Review by fleupold — APPROVED** [review]
Looks great!

It would be awesome if you could address the remaining comments before merging.

File: test/merkle.js (hunk 6 above)
Comment (bh2smith) [reply test/merkle.js]: I was not aware I could "choose" not to import things I don't need from the utilities file. These are leftovers from cut contract (I believe). Not only shall I remove the import statements, but will get rid of any functions that are not needed anywhere in the next commit.

File: test/merkle.js (hunk 8 above)
Comment (bh2smith) [reply test/merkle.js]: I would say that it doesn't really add any value as the height 2 tree is non-trivial anyway. However, while experimenting with this, I did notice that our checkMembership function is not  equip to handle any lop-sided trees (i.e. those without an exact power of 2 number of leaves) that are not implemented with 'padded' zeros. Will remove this until further notice.

File: test/utilities.js (hunk 9 above)
Comment (bh2smith) [reply test/utilities.js:8]: I will take a look into it and really hope to find something.

File: test/plasma.js (hunk 10 above)
Comment (bh2smith) [reply test/plasma.js]: They aren't, but I didn't want to make a whole new one just for this.

File: test/merkle.js (hunk 11 above)
Comment (bh2smith) [reply test/merkle.js]: Ahhh, Yes. Of course.

Comment (bh2smith) [reply test/merkle.js]: Although, I have removed the second section of tests for this contract. In this case it simply boils down to moving the before-async stuff into the one block of tests. 

PS - I wasn't able to take the await out of the async because of some js-error that I am too young to understand.

```
SyntaxError: await is only valid in async function
```

File: test/utilities.js (hunk 9 above)
Comment (bh2smith) [reply test/utilities.js:8]: I looked around briefly and will continue to look, but I think I will add a "good first issue" to investigate this.

File: test/plasma.js (hunk 10 above)
Comment (fleupold) [reply test/plasma.js]: I understand. Usually it's nice to keep changes isolated in case we have to revert certain pull requests (then we really only revert what the pull request was about).
Doing small cleanups like this are likely going to be immediately accepted and merged so it's really just the overhead of creating another pr. I wonder if we can make our git workflow more efficient to encourage that.

I usually `git stash` changes like that on the unrelated branch and once I'm done with the current PR, I rebase back on master, un-stash them and create a new branch/commit with them.
